### PR TITLE
OBSTACLE_DISTANCE: add higher angle resolution and angle offset

### PR DESCRIPTION
**Pull request by magicrub**
This MAVLink v2 packet allows for flexible obstacle detection from multiple sensors of varying resolution while still being backwards compatible.

increment_f (to override increment):
When forced to send exactly 72 samples, having an integer spacing just didn't make sense. Needed to be float so that the correct angular range was understood

angle_offset:
Same as, and adds to, param "[PRX_YAW_CORR](http://ardupilot.org/rover/docs/parameters.html#prx-yaw-corr-proximity-sensor-yaw-correction)". However, if you have two sensors you can send multiple messages covering different areas (at varying resolutions). For example, a 360 degree spinning lidar + long-range forward radar/lidar + wide angle sonar(s). With multiple messages, we can aggregate a better picture and down-scale the resolution to whatever the flight controller can handle in that version. And that will improve as we get fancier flight controllers with more memory Meanwhile, the data is no longer limited by the comms link, just on how to store it once received.

count:
Allow the option to send less than 72 distance values.

### CODE REVIEW

File: message_definitions/v1.0/common.xml [hunk 1]
```diff
@@ -4552,6 +4552,10 @@
       <field type="uint8_t" name="increment" units="deg">Angular width in degrees of each array element.</field>
       <field type="uint16_t" name="min_distance" units="cm">Minimum distance the sensor can measure.</field>
       <field type="uint16_t" name="max_distance" units="cm">Maximum distance the sensor can measure.</field>
+      <extensions/>
+      <field type="float" name="increment_f" units="deg">Angular width in degrees of each array element as a float. If >0 then this value is used instead of uint8_t incremenet.</field>
```
Comment (OXINARF) [inline message_definitions/v1.0/common.xml]: incremenet -> increment

File: message_definitions/v1.0/common.xml [hunk 2]
```diff
@@ -4552,6 +4552,10 @@
...
+      <field type="float" name="angle_offset" units="deg">Relative angle offset of first distance with index 0. Value of 0 corresponding to local North/forward and matches existing map and otherwise adds an offset. Positive is to the right.</field>
+      <field type="uint8_t" name="count">Count of distances values. Valid range is 1-72. A value of 0 or >72 is treated as 72.</field>
```
Comment (OXINARF) [inline message_definitions/v1.0/common.xml]: This isn't needed, it is already explicitly said in the distances array whether a distance is valid, invalid or unknown; it's already possible to send less than 72 samples.

Comment (magicrub) [reply message_definitions/v1.0/common.xml]: It solves the vague problem where you have data, then invalid, then data, then invalid, invalid... you don't know where the end is. This is to more easily allow for bad samples in variable sized packets from mixed sensors.

**Comment by magicrub** [comment]
> I have my doubts about the real-life usefulness of increment_f

Very important. @rmackay9 and @jmachuca77 really want it too. The issue is if you have 72 samples with integer degree, what happens if those 72 samples cover 90 degrees? Or 20 degrees? This is to prepare the Proximity library for higher resolution target tracking which is being worked on now..

File: message_definitions/v1.0/common.xml (hunk 2 above)
Comment (WickedShell) [reply message_definitions/v1.0/common.xml]: How does this create variable sized packets? Extensions are always added at the end, so any of these being set means none of the array will 0 compressed, so no bandwidth is saved,

Comment (magicrub) [reply message_definitions/v1.0/common.xml]: This has nothing to do with MAVLink 2 packet length truncation. The problem is 0 is a valid distance so if you're using the packet correctly then unused values don't save bandwidth anyway. This allows the receiver to know there are less than 72 samples while allowing the UINT16_MAX sample(s) to signify that they still take up space on the arc.

Comment (WickedShell) [reply message_definitions/v1.0/common.xml]: The iteration time for 72 items is really cheap on this though, so it's not saving a lot of time, and does add to the transmission overhead. (The 0 == 72 logic is also a bit painful if this is retained)

Comment (magicrub) [reply message_definitions/v1.0/common.xml]: example of a sensor that populates 5 values without the counts variable, but we don't know that it's only ever populating 5 values and the rest are always UINT16_MAX.

(treated as 5 values)
1,2,3,4,5,UINT16_MAX.....

(treated as 5 values, if we assume last valid sample is last index)
1,UINT16_MAX,3,4,5,UINT16_MAX..... (index 2 has invalid/unknown sample)

(treated as 1 sample and we lose the knowledge that there are 4 more zones of invalid data)
1,UINT16_MAX,UINT16_MAX,UINT16_MAX,UINT16_MAX,UINT16_MAX..... (index 2,3,4,5 have invalid/unknown sample).

File: message_definitions/v1.0/common.xml (hunk 1 above)
Comment (magicrub) [reply message_definitions/v1.0/common.xml]: fixed

File: message_definitions/v1.0/common.xml (hunk 2 above)
Comment (magicrub) [reply message_definitions/v1.0/common.xml]: I talked it over with @jmachuca77 and, even though this help would create less work on the autopilot (receiving) side athe cost of just one transport byte, the end result distance mapping would likely always be the same. So, I'll remove COUNT entry. All Hail STM32F7!

**Comment by magicrub** [comment]
Here's the upstream PR https://github.com/mavlink/mavlink/pull/1142

**Comment by magicrub** [comment]
This has been [merged into upstream](https://github.com/mavlink/mavlink/pull/1142). I've replaced the commit on this PR branch to be a cherry-pick from upstream. I'll merge this once it passes travis.

**Pull request merged**